HW0/P5/P5.txt

```diff
@@ -1 +1,46 @@
-# Your answers here.
+# P5 Answers
+# Stephen Fang
+# shifang@g.harvard.edu / stephenshifang@gmail.com
+
+1. 
+255 seconds.
+
+2. 
+Split the bags evenly among 8 cashiers.
+Time needed = 256/8 - 1 = 31 seconds
+Split the 8 subtotals evenly between 4 cashiers.
+Time needed = 8/4 - 1 = 1 second
+Split the 4 subtotals evenly between 2 cashiers.
+Time needed = 4/2 - 1 = 1 second
+Finally, one cashier can add the 2 subtotals to get the total.
+Time needed = 1 second
+Total time spent = 31 + 1 + 1 + 1 = 34 seconds
+
+3.
+Split the bags evenly among 128 cashiers and generate 128 subtotals.
+Time needed = 1 second.
+Split the subtotals evenly among 64 cashiers and generate 64 subtotals.
+Time needed = 1 second.
+And recurse until we have 2 subtotals left.
+Total time needed = log2(256) = 8 seconds
+
+4.
+If we keep splitting the bags so that each cashier only needs to add 2 numbers during each iteration, total time needed = ceiling(log2(N)).
+
+5.
+Considering overhead associated with people communicating, 
+each split and count would need 1 second of adding and 1 second of communication overhead except for the final count.
+So total time needed = log2(256)*2-1 = 15 seconds
+
+6.
+If each cashier comes to me and takes 1 bag, then it will take 256 seconds to distribute the bags.
```

Comment:
I believe that you can obtain a faster result by allowing the cashiers to start summing up their totals while the first person is still distributing the bags.  That way, some of the log-2(N) steps can occur in parallel with the initial N distribution steps.  By my calculations, that brings the total for the first part down from 272 to 260 seconds (not a huge improvement, but everything helps!).